### PR TITLE
(feat) ComponentEvents interface

### PR DESCRIPTION
**Pull request by dummdidumm**
This adds the possibility to use a reserved interface name `ComponentEvents` and define all possible events within it.
Also adds autocompletion for these events.
Also disables autocompletions from HTMLPlugin on component tags.
NO support for generics.

#424 #304 #442 

TODO:
- [x] tests
- [x] when interface is defined, make events strict so that events that do not exist are marked as error
- [x] docs

### CODE REVIEW

File: packages/svelte2tsx/src/nodes/ComponentEvents.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        const map = new Map<string, { type: string; doc?: string }>();
+
+        node.members.filter(ts.isPropertySignature).forEach((member) => {
+            map.set(member.name.getText(), {
+                type: member.type?.getText() || 'Event',
+                doc: this.getDoc(node, member),
+            });
+        });
+
+        return map;
```
Comment (jasonlyu123) [inline packages/svelte2tsx/src/nodes/ComponentEvents.ts:41]: The name here is possibly a string literal or a computed property. i.e.
```
interface ComponentEvents {
    'kebab-case': CustomEvent<number>
    [SOME_CONSTANT]: CustomEvent<number>
}
```
`getText` in either situation would be wrong. if it's a string literal the text can be extracted from `member.name.text`.
About the computed property, from my testing, computed event handler like <Component on:[SOME_CONSTANT] /> doesn't work. Maybe we could make it works here but filter out in the completion. Or simply doesn't support it as it doesn't make much sense when you can't use it in markup.

**Review by jasonlyu123 — CHANGES_REQUESTED** [review]
Mostly great, only an issue about string literal or computed property name needs more works

File: packages/language-server/src/plugins/typescript/features/CompletionProvider.ts [hunk 2]
```diff
@@ -71,10 +72,11 @@ export class CompletionsProviderImpl implements CompletionsProvider<CompletionEn
             ? triggerCharacter
             : undefined;
         const isCustomTriggerCharacter = triggerKind === CompletionTriggerKind.TriggerCharacter;
+        const isInvoked = triggerKind === CompletionTriggerKind.Invoked;
 
         // ignore any custom trigger character specified in server capabilities
         //  and is not allow by ts
-        if (isCustomTriggerCharacter && !validTriggerCharacter) {
+        if (isCustomTriggerCharacter && !validTriggerCharacter && !isInvoked) {
```
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/language-server/src/plugins/typescript/features/CompletionProvider.ts:79]: No, the added check makes it possible to get completions when doing Ctrl+Space (in my keybindings). I'll update the comment above.

File: packages/svelte2tsx/src/nodes/ComponentEvents.ts (hunk 1 above)
Comment (dummdidumm) [reply packages/svelte2tsx/src/nodes/ComponentEvents.ts:41]: Good catch! We should definitely support string literals. About the constant, we could go to the root and check if it's defined there and get the value from there - so we at least support it a little. I will also add docs about this.

File: packages/language-server/src/plugins/typescript/features/CompletionProvider.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasonlyu123) [reply packages/language-server/src/plugins/typescript/features/CompletionProvider.ts:79]: What did I mean it that if `isCustomTriggerCharacter` is true isn't `isInvoked` always be false?

File: packages/svelte2tsx/src/svelte2tsx.ts [hunk 3]
```diff
@@ -1016,6 +1032,7 @@ export function svelte2tsx(
         str,
         uses$$props || uses$$restProps,
         !!options?.strictMode,
+        events instanceof ComponentEventsFromInterface,
```
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/svelte2tsx/src/svelte2tsx.ts:1035]: Yeah makes sense, `addComponentExport` has way too many parameters right now 😄

File: packages/language-server/src/plugins/typescript/features/CompletionProvider.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/language-server/src/plugins/typescript/features/CompletionProvider.ts:79]: Oh yes, you are completely right ... why did I do this? 😄 I'll remove it.

File: packages/svelte2tsx/test/svelte2tsx/index.js [hunk 4]
```diff
@@ -16,15 +16,20 @@ describe('svelte2tsx', () => {
...
-            const { map, code} = svelte2tsx(input, {
+            const output = svelte2tsx(input, {
 				strictMode: dir.includes('strictMode'),
 				isTsFile: dir.startsWith('ts-'),
 				filename: 'input.svelte'
 			});
-            assert.equal(code, expectedOutput);
+			assert.equal(output.code, expectedOutput);
+			if (expecterOtherOutput) {
+				expecterOtherOutput(output);
```
Comment (dummdidumm) [inline packages/svelte2tsx/test/svelte2tsx/index.js:31]: I enhanced the test suite so you can test other outputs of `svelte2tsx`, see `component-events-interface-constant/expected.js` for an example.

File: packages/svelte2tsx/src/nodes/ComponentEvents.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    private throwError(prop: ts.PropertyName) {
+        const error: any = new Error(
+            'The ComponentEvents interface can only have properties of type ' +
+                'Identifier, StringLiteral or ComputedPropertyName. ' +
+                'In case of ComputedPropertyName, ' +
+                'it must be a const declared within the component and initialized with a string.',
+        );
+        error.start = toLineColumn(prop.getStart());
+        error.end = toLineColumn(prop.getEnd());
+        throw error;
```
Comment (dummdidumm) [inline packages/svelte2tsx/src/nodes/ComponentEvents.ts:80]: This throws if the interface definition does not match our expextations. I hope this makes it clear to the user what is possible and what is not.

File: packages/svelte2tsx/src/nodes/ComponentEvents.ts [hunk 6]
```diff
@@ -41,6 +42,52 @@ export class ComponentEventsFromInterface extends ComponentEvents {
         return map;
     }
 
+    private getName(prop: ts.PropertyName) {
+        if (ts.isIdentifier(prop) || ts.isStringLiteral(prop)) {
```
Comment (jasonlyu123) [inline packages/svelte2tsx/src/nodes/ComponentEvents.ts:46]: Also NumberLiteral

Comment (dummdidumm) [reply packages/svelte2tsx/src/nodes/ComponentEvents.ts:46]: According to the Svelte typings, only `string` is possible for the events: https://github.com/sveltejs/svelte/blob/master/src/runtime/internal/lifecycle.ts#L32

Comment (jasonlyu123) [reply packages/svelte2tsx/src/nodes/ComponentEvents.ts:46]: If it's official now then is fine. I just tried it in REPL and find it possible as a number XD.

Comment (dummdidumm) [reply packages/svelte2tsx/src/nodes/ComponentEvents.ts:46]: I think it will be stringified, so the correct typing would be to only allow `string`.

**Comment by dummdidumm** [comment]
Thanks for the feedback! I added the requested changes. Let me know if it's all good now for you.

**Review by jasonlyu123 — APPROVED** [review]

**Comment by nateshmbhat** [comment]
can someone please provide on how to define type defs to include events too ?

**Comment by dummdidumm** [comment]
This is not official yet until [this RFC](https://github.com/sveltejs/rfcs/pull/38) is finalized and merged. You can type `createEventDispatcher` though.

**Pull request merged**